File: erpnext/stock/page/serial_attributes/serial_attributes.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

"""Backend for the «Serial Attributes» desk page.

Lists Serial Nos with their `Additional Attribute Row` values, and applies attribute values
to many serials at once — either the ones ticked in the table or a pasted list of serial
numbers.
"""

import frappe
from frappe import _
from frappe.utils import cint

ROW_DOCTYPE = "Additional Attribute Row"
ROW_FIELDNAME = "additional_attributes"
MAX_ROWS = 2000


def _check_read():
	frappe.has_permission("Serial No", "read", throw=True)


def _check_write():
	frappe.has_permission("Serial No", "write", throw=True)


@frappe.whitelist()
def get_serials(
	item_code=None,
	item_group=None,
	warehouse=None,
	status=None,
	attribute=None,
	value=None,
	missing_only=0,
	in_stock_only=1,
	limit=500,
):
	"""Serial Nos matching the filters, each with its additional attributes.

	`missing_only` flips the `attribute` filter into "serials that have no row for this
	attribute" — the list you then select and fill in.
	"""
	_check_read()

	missing_only = cint(missing_only)
	in_stock_only = cint(in_stock_only)
	limit = min(cint(limit) or 500, MAX_ROWS)

	sn = frappe.qb.DocType("Serial No")
	query = (
		frappe.qb.from_(sn)
		.select(sn.name, sn.item_code, sn.item_name, sn.warehouse, sn.status)
		.orderby(sn.item_code)
		.orderby(sn.name)
		.limit(limit)
	)

	if item_code:
		query = query.where(sn.item_code == item_code)
	if warehouse:
		query = query.where(sn.warehouse == warehouse)
	if status:
		query = query.where(sn.status == status)
	elif in_stock_only:
		query = query.where(sn.warehouse.notnull() & (sn.warehouse != ""))

	if item_group:
		item = frappe.qb.DocType("Item")
		query = query.join(item).on(item.name == sn.item_code).where(item.item_group == item_group)

	if attribute:
		row = frappe.qb.DocType(ROW_DOCTYPE)
		matching = (
			frappe.qb.from_(row)
			.select(row.parent)
			.where((row.parenttype == "Serial No") & (row.attribute == attribute))
		)
		if value:
			matching = matching.where(row.value == value)

		if missing_only:
			query = query.where(sn.name.notin(matching))
		else:
			query = query.where(sn.name.isin(matching))

	serials = query.run(as_dict=True)
	if not serials:
		return {"serials": [], "attributes": _column_attributes(attribute)}

	names = [d.name for d in serials]
	rows = frappe.get_all(
		ROW_DOCTYPE,
		filters={"parenttype": "Serial No", "parent": ("in", names)},
		fields=["parent", "attribute", "value", "notes"],
		order_by="idx",
	)

	value_labels = _value_labels([d.value for d in rows])
	by_serial = {}
	for row in rows:
		by_serial.setdefault(row.parent, {})[row.attribute] = {
			"value": row.value,
			"label": value_labels.get(row.value, row.value),
			"notes": row.notes,
		}

	for serial in serials:
		serial["attributes"] = by_serial.get(serial.name, {})

	return {"serials": serials, "attributes": _column_attributes(attribute)}


def _column_attributes(attribute=None):
	"""Which attributes get a column: the filtered one, or every enabled attribute."""
	if attribute:
		return [attribute]

	return [
		d.name
		for d in frappe.get_all(
			"Additional Attribute", filters={"disabled": 0}, fields=["name"], order_by="attribute_name"
		)
	]


def _value_labels(value_names):
	value_names = [d for d in set(value_names) if d]
	if not value_names:
		return {}

	return {
		d.name: d.value
		for d in frappe.get_all(
			"Additional Attribute Value", filters={"name": ("in", value_names)}, fields=["name", "value"]
		)
	}


@frappe.whitelist()
def parse_serials(text, item_code=None):
	"""Split a pasted blob into known / unknown serial numbers.

	Accepts newlines, commas, semicolons, tabs and spaces as separators.
	"""
	_check_read()

	tokens = []
	for chunk in (text or "").replace(",", "\n").replace(";", "\n").replace("\t", "\n").split("\n"):
		token = chunk.strip()
		if token:
			tokens.append(token)

	# preserve the pasted order, drop repeats
	seen = set()
	ordered = []
	for token in tokens:
		if token not in seen:
			seen.add(token)
			ordered.append(token)

	if not ordered:
		return {"found": [], "unknown": [], "wrong_item": []}

	filters = {"name": ("in", ordered)}
	existing = {
		d.name: d.item_code
		for d in frappe.get_all("Serial No", filters=filters, fields=["name", "item_code"])
	}

	found, unknown, wrong_item = [], [], []
	for token in ordered:
		if token not in existing:
			unknown.append(token)
		elif item_code and existing[token] != item_code:
			wrong_item.append(token)
		else:
			found.append(token)

	return {"found": found, "unknown": unknown, "wrong_item": wrong_item}


@frappe.whitelist()
def set_attribute(serials, attribute, value=None, notes=None, overwrite=1, clear=0):
	"""Set (or clear) one attribute on many Serial Nos.

	`overwrite=0` skips serials that already carry a value for this attribute — the safe mode
	for filling gaps. `clear=1` removes the attribute row instead of setting it.
	"""
	_check_write()

	serials = frappe.parse_json(serials) if isinstance(serials, str) else serials
	serials = [d for d in (serials or []) if d]
	if not serials:
		frappe.throw(_("No serial numbers selected"))

	overwrite = cint(overwrite)
	clear = cint(clear)

	if not attribute:
		frappe.throw(_("Attribute is required"))
	if not clear and not value:
		frappe.throw(_("Value is required"))

	if value:
		value_attribute = frappe.db.get_value("Additional Attribute Value", value, "attribute")
		if not value_attribute:
			frappe.throw(_("Additional Attribute Value {0} does not exist").format(frappe.bold(value)))
		if value_attribute != attribute:
			frappe.throw(
				_("Value {0} belongs to attribute {1}, not {2}").format(
					frappe.bold(value), frappe.bold(value_attribute), frappe.bold(attribute)
				)
			)

	updated, skipped, failed = [], [], []
	for name in serials:
		# one savepoint per serial: a bad row must not undo the serials already written
		save_point = "serial_attributes"
		frappe.db.savepoint(save_point)
		try:
			doc = frappe.get_doc("Serial No", name)
			if not doc.meta.get_field(ROW_FIELDNAME):
				frappe.throw(_("Serial No has no Additional Attributes table"))

			existing = [d for d in doc.get(ROW_FIELDNAME) if d.attribute == attribute]
			if existing and not overwrite and not clear:
				skipped.append(name)
				continue

			for row in existing:
				doc.remove(row)

			if not clear:
				doc.append(ROW_FIELDNAME, {"attribute": attribute, "value": value, "notes": notes})

			doc.save()
			updated.append(name)
		except Exception as e:
			frappe.db.rollback(save_point=save_point)
			failed.append({"serial": name, "error": str(e)})

	frappe.db.commit()

	return {"updated": updated, "skipped": skipped, "failed": failed}

File: erpnext/stock/page/serial_attributes/serial_attributes.js
frappe.pages["serial-attributes"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Serial Attributes"),
		single_column: true,
	});

	const METHOD = "erpnext.stock.page.serial_attributes.serial_attributes";

	let rows = [];
	let columns = [];
	const selected = new Set();

	const item_field = page.add_field({
		label: __("Item"),
		fieldtype: "Link",
		fieldname: "item_code",
		options: "Item",
		change: () => refresh(),
	});

	const item_group_field = page.add_field({
		label: __("Item Group"),
		fieldtype: "Link",
		fieldname: "item_group",
		options: "Item Group",
		change: () => refresh(),
	});

	const warehouse_field = page.add_field({
		label: __("Warehouse"),
		fieldtype: "Link",
		fieldname: "warehouse",
		options: "Warehouse",
		change: () => refresh(),
	});

	const status_field = page.add_field({
		label: __("Status"),
		fieldtype: "Select",
		fieldname: "status",
		options: ["", "Active", "Inactive", "Consumed", "Delivered", "Expired"],
		change: () => refresh(),
	});

	const attribute_field = page.add_field({
		label: __("Attribute"),
		fieldtype: "Link",
		fieldname: "attribute",
		options: "Additional Attribute",
		change: () => {
			value_field.set_value("");
			refresh();
		},
	});

	const value_field = page.add_field({
		label: __("Value"),
		fieldtype: "Link",
		fieldname: "value",
		options: "Additional Attribute Value",
		get_query: () => ({ filters: { attribute: attribute_field.get_value() || undefined } }),
		change: () => refresh(),
	});

	const missing_field = page.add_field({
		label: __("Only Missing This Attribute"),
		fieldtype: "Check",
		fieldname: "missing_only",
		change: () => refresh(),
	});

	page.set_primary_action(__("Set Attribute"), () => set_dialog([...selected]), "edit");
	page.add_menu_item(__("Set By Pasted Serial Numbers"), () => paste_dialog());
	page.add_menu_item(__("Reload"), () => refresh());

	const $body = $(`
		<div class="serial-attributes">
			<div class="sa-toolbar text-muted small"></div>
			<div class="sa-table"></div>
		</div>
	`).appendTo(page.body);

	$("<style>")
		.text(
			`
		.serial-attributes .sa-toolbar { margin: 6px 0 10px; }
		.serial-attributes table { font-size: 13px; width: 100%; }
		.serial-attributes th, .serial-attributes td { padding: 6px 8px; vertical-align: middle; }
		.serial-attributes td.sa-attr { cursor: pointer; }
		.serial-attributes td.sa-attr:hover { background: var(--fg-hover-color); }
		.serial-attributes td.sa-empty { color: var(--text-muted); font-style: italic; }
		.serial-attributes tr.sa-checked { background: var(--highlight-color); }
		.serial-attributes .sa-check { cursor: pointer; }
	`
		)
		.appendTo("head");

	function refresh() {
		frappe.call({
			method: `${METHOD}.get_serials`,
			args: {
				item_code: item_field.get_value(),
				item_group: item_group_field.get_value(),
				warehouse: warehouse_field.get_value(),
				status: status_field.get_value(),
				attribute: attribute_field.get_value(),
				value: value_field.get_value(),
				missing_only: missing_field.get_value() ? 1 : 0,
			},
			callback: (r) => {
				const data = r.message || {};
				rows = data.serials || [];
				columns = data.attributes || [];
				selected.clear();
				render();
			},
		});
	}

	function render() {
		if (!rows.length) {
			$body
				.find(".sa-table")
				.html(`<p class="text-muted">${__("No serial numbers match these filters")}</p>`);
			update_toolbar();
			return;
		}

		const head = columns.map((c) => `<th>${frappe.utils.escape_html(c)}</th>`).join("");

		const body = rows
			.map((row) => {
				const cells = columns
					.map((attribute) => {
						const attr = (row.attributes || {})[attribute];
						const label = attr
							? frappe.utils.escape_html(attr.label || attr.value)
							: __("Not set");
						const title = attr && attr.notes ? frappe.utils.escape_html(attr.notes) : "";
						return `<td class="sa-attr ${attr ? "" : "sa-empty"}"
								data-serial="${frappe.utils.escape_html(row.name)}"
								data-attribute="${frappe.utils.escape_html(attribute)}"
								title="${title}">${label}</td>`;
					})
					.join("");

				return `
					<tr data-serial="${frappe.utils.escape_html(row.name)}">
						<td><input type="checkbox" class="sa-check" data-serial="${frappe.utils.escape_html(row.name)}"></td>
						<td><a href="/app/serial-no/${encodeURIComponent(row.name)}">${frappe.utils.escape_html(row.name)}</a></td>
						<td>${frappe.utils.escape_html(row.item_code || "")}</td>
						<td>${frappe.utils.escape_html(row.warehouse || "")}</td>
						<td>${frappe.utils.escape_html(row.status || "")}</td>
						${cells}
					</tr>`;
			})
			.join("");

		$body.find(".sa-table").html(`
			<table class="table table-bordered">
				<thead>
					<tr>
						<th style="width: 30px"><input type="checkbox" class="sa-check-all"></th>
						<th>${__("Serial No")}</th>
						<th>${__("Item")}</th>
						<th>${__("Warehouse")}</th>
						<th>${__("Status")}</th>
						${head}
					</tr>
				</thead>
				<tbody>${body}</tbody>
			</table>
		`);

		$body.find(".sa-check-all").on("change", function () {
			const checked = this.checked;
			$body.find(".sa-check").prop("checked", checked);
			selected.clear();
			if (checked) rows.forEach((row) => selected.add(row.name));
			$body.find("tbody tr").toggleClass("sa-checked", checked);
			update_toolbar();
		});

		$body.find(".sa-check").on("change", function () {
			const serial = $(this).data("serial");
			if (this.checked) selected.add(serial);
			else selected.delete(serial);
			$(this).closest("tr").toggleClass("sa-checked", this.checked);
			update_toolbar();
		});

		$body.find("td.sa-attr").on("click", function () {
			set_dialog([$(this).data("serial")], $(this).data("attribute"));
		});

		update_toolbar();
	}

	function update_toolbar() {
		$body
			.find(".sa-toolbar")
			.html(`${__("Shown: {0}", [rows.length])} &nbsp;·&nbsp; ${__("Selected: {0}", [selected.size])}`);
	}

	function set_dialog(serials, attribute) {
		if (!serials || !serials.length) {
			frappe.msgprint(__("Select at least one serial number"));
			return;
		}

		const dialog = new frappe.ui.Dialog({
			title: __("Set Attribute for {0} Serial No(s)", [serials.length]),
			fields: [
				{
					label: __("Attribute"),
					fieldname: "attribute",
					fieldtype: "Link",
					options: "Additional Attribute",
					reqd: 1,
					default: attribute || attribute_field.get_value(),
					onchange: () => dialog.set_value("value", ""),
				},
				{
					label: __("Value"),
					fieldname: "value",
					fieldtype: "Link",
					options: "Additional Attribute Value",
					depends_on: "eval:!doc.clear",
					get_query: () => ({ filters: { attribute: dialog.get_value("attribute") || undefined } }),
				},
				{ fieldtype: "Column Break" },
				{
					label: __("Overwrite Existing Value"),
					fieldname: "overwrite",
					fieldtype: "Check",
					default: 1,
					description: __("Uncheck to fill only serials that have no value yet"),
				},
				{
					label: __("Clear Attribute"),
					fieldname: "clear",
					fieldtype: "Check",
					description: __("Remove this attribute from the selected serial numbers"),
				},
				{ fieldtype: "Section Break" },
				{ label: __("Notes"), fieldname: "notes", fieldtype: "Small Text" },
			],
			primary_action_label: __("Apply"),
			primary_action: (values) => {
				frappe.call({
					method: `${METHOD}.set_attribute`,
					args: {
						serials: serials,
						attribute: values.attribute,
						value: values.value,
						notes: values.notes,
						overwrite: values.overwrite ? 1 : 0,
						clear: values.clear ? 1 : 0,
					},
					freeze: true,
					freeze_message: __("Updating serial numbers..."),
					callback: (r) => {
						dialog.hide();
						show_result(r.message || {});
						refresh();
					},
				});
			},
		});

		dialog.show();
	}

	function show_result(result) {
		const parts = [__("Updated: {0}", [(result.updated || []).length])];
		if ((result.skipped || []).length) {
			parts.push(__("Skipped (already set): {0}", [result.skipped.length]));
		}
		if ((result.failed || []).length) {
			parts.push(__("Failed: {0}", [result.failed.length]));
			parts.push(
				result.failed
					.slice(0, 10)
					.map((f) => `${frappe.utils.escape_html(f.serial)}: ${frappe.utils.escape_html(f.error)}`)
					.join("<br>")
			);
		}
		frappe.msgprint({ title: __("Bulk Update"), message: parts.join("<br>"), indicator: "blue" });
	}

	function paste_dialog() {
		const dialog = new frappe.ui.Dialog({
			title: __("Set By Pasted Serial Numbers"),
			fields: [
				{
					label: __("Serial Numbers"),
					fieldname: "serials",
					fieldtype: "Small Text",
					reqd: 1,
					description: __("Separate by new line, comma, semicolon, tab or space"),
				},
			],
			primary_action_label: __("Check List"),
			primary_action: (values) => {
				frappe.call({
					method: `${METHOD}.parse_serials`,
					args: { text: values.serials, item_code: item_field.get_value() },
					callback: (r) => {
						const parsed = r.message || {};
						const found = parsed.found || [];
						const unknown = parsed.unknown || [];
						const wrong_item = parsed.wrong_item || [];

						if (unknown.length || wrong_item.length) {
							const lines = [];
							if (unknown.length) {
								lines.push(
									`<b>${__("Unknown serial numbers: {0}", [unknown.length])}</b><br>` +
										frappe.utils.escape_html(unknown.slice(0, 20).join(", "))
								);
							}
							if (wrong_item.length) {
								lines.push(
									`<b>${__("Belong to another Item: {0}", [wrong_item.length])}</b><br>` +
										frappe.utils.escape_html(wrong_item.slice(0, 20).join(", "))
								);
							}
							frappe.msgprint({
								title: __("Some Serial Numbers Were Skipped"),
								message: lines.join("<br><br>"),
								indicator: "orange",
							});
						}

						if (!found.length) {
							return;
						}

						dialog.hide();
						set_dialog(found);
					},
				});
			},
		});

		dialog.show();
	}

	refresh();
};
